Improve Dependency Resolution for Workflow Optimization

oliverqx opened the pull request:
<!--
Please fill in each sections of this template, and delete any sections that are not relevant.
-->
/claim #2235

**Description**
This PR implements module package-based dependency resolution as requested by issue #2235. 

**Additional context**
Linked issue explains it quite well.

**Metadata**
- Fixes #2235


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
  - Enhanced the `Git Changes Action` to export a variable with changed Go modules in pull requests, introducing `Module to module` and `Module to package` dependency resolution methods.
  - Customization options include `levelOfDependencyResolution` for specifying resolution method and examples for testing based on modified modules and dependencies.
- **Refactor**
  - Updated function calls in the main function to include an additional argument for specifying the type of dependency.
- **Documentation**
  - README.md now provides details on new features, dependency resolution methods, and examples for testing with different resolution levels.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

trajan0x commented:
@oliverqx btw working on some changes to let these wokflows run correctly

oliverqx commented:
@trajan0x I created this [repository](https://github.com/oliverqx/sanguine-git-changes-action) (which is a synapse/sanguine clone without all the git history and without unrelated github workflows) where I've been running the github action. Would be nice to have the action tested there to avoid polluting this repository. I'd like to open some PRs as test cases so I've been reading the repo and I'm struggling to find the following case:

```
ModuleA/PackageA -> ModuleB/PackageB -> ModuleC/PackageC
```
> where -> represents  'depends on'

Conditions:
- ModuleA/PackageA should not import ModuleC/PackageC directly

Once a case with those conditions is found, the plan is to open two PRs:
1. Change ModuleC/PackageC and have the action flag ModuleA and ModuleB as modified.
2. Change ModuleC/PackageA and prove action does not flag ModuleA and ModuleB as modified.

Do you by any change know of such a relationship within the codebase?

oliverqx commented on contrib/git-changes-action/detector/detector.go:
```diff
@@ -41,28 +41,62 @@
+
+      modules[module.Path] = changed
+    }
+  }
```
will resolve conversation as this code is not the most recent commit

oliverqx replied:
can you review the most recent changes?

oliverqx commented on contrib/git-changes-action/detector/detector.go:
```diff
@@ -19,16 +19,43 @@ import (
+
+  packages[packageName] = changed
+  return changed
+}
```
will resolve conversation as this code is not the most recent commit

oliverqx commented on contrib/git-changes-action/detector/detector.go:
```diff
@@ -41,28 +41,69 @@
+
+      modules[module.Path] = changed
+    }
+  }
```
took the advice, will resolve conversation

trajan0x commented:
sorry @oliverqx seeing this isnt' a draft anymore. Is this ready for revie?

oliverqx commented:
@trajan0x It is..  about the structure of my solution. I added a parameter to the detector function to maintain module level dependency resolution as an option. This creates two logic branches to all subsequent functions, and its a bit strange. I was thinking I could create two subpackages in detector:

1. moduledetector: will contain function, utilities and tests to detect changed modules on module level deps
2. packagedetector: will contain function, utilities and tests, to detect changed modules on package level deps
3. detector: will contain git related functions (getchangetree, gethead, etc)

This fixes a lot of the lint issues and, I think, its a better dx.  I've already done it locally but since its quite a structural change, I wanted to ask if thats a good route to take before I push the changes.

trajan0x commented:
@oliverqx think this makes a lot of sense

oliverqx commented:
@trajan0x  just pushed the changes, i think its ready to be reviewed

trajan0x commented:
fyi, available to try no w/ docker tag ghcr.io/synapsecns/sanguine/git-changes-action:07a4696bbbafbba2d3ad7ee828b4c8adb21b4d20

oliverqx commented:
I created this [repo](https://github.com/oliverqx/sanguine-git-changes-action). Its a clone of sanguine, without the git history. I removed all github workflows and added a very simple workflow that uses the image.

I also opened [4 PRs](https://github.com/oliverqx/sanguine-git-changes-action/pulls) and they all return modules modified based on package dependencies.

Its a bit difficult to follow as sanguine is quite complex and the dependencies can run deep, so I created this other repository, that is easier to follow as its a smaller example. PRs also explain package dependencies quite well, starting here: https://github.com/oliverqx/showcase/pull/1

The test should also pass.

trajan0x commented:
@oliverqx would you mind running go `generate ./...`?

Also going to merge this in #2391 just to make sure nothing breaks, then can gtm

oliverqx commented:
Just ran go generate, looks like the solution will be accepted so I'll update documentation.

trajan0x commented:
Note to self: would like to investigate why 89d9aa9e2ca3ada2f10e89fb69d2e393e6624b6d triggered `contrib/promexporter` as I don't see how that dependency would've been triggered

oliverqx commented:
> Note to self: would like to investigate why [89d9aa9](https://github.com/synapsecns/sanguine/commit/89d9aa9e2ca3ada2f10e89fb69d2e393e6624b6d) triggered `contrib/promexporter` as I don't see how that dependency would've been triggered

Looking into this, quite a strange bug

oliverqx commented:
Would you let me know if the README.md is accurate and understandable

oliverqx commented:
@trajan0x today im pretty much free, will be looking into that bug. I think Its got to do with ct.HasPath() might need to do some edits there, will keep you updated

trajan0x commented:
Appreciate it

oliverqx commented:
@trajan0x Actually I realized its not a bug. 

I created this PR in my own sanguine replica: https://github.com/oliverqx/sanguine-git-changes/pull/4

And I added an fmt.Println to all ./contrib/promexporter dependencies to my [image](https://github.com/users/oliverqx/packages/container/package/git-changes-action) and it looks like ./ethergo/submitter is one of the dependencies. Keep in mind that this array a contains all transitive dependencies, so manually following it might be quite complicated. 

![Screenshot 2024-04-02 at 16 44 41](https://github.com/synapsecns/sanguine/assets/8559757/f3ad5a9a-84b2-4a20-b259-924c06064ec5)

oliverqx commented:
@trajan0x just checking in, I think the solution is ready. Please let me know is there's anything else I could do

trajan0x commented:
Okay, will stage the latest build today

trajan0x commented:
LGTM

oliverqx commented:
sounds good, thanks!

oliverqx commented:
@trajan0x once the bounty has been settled, I can provide support if anything is wrong. I know my solution is right, but nonetheless. 

Was also wondering if I could get a linkedin review or a recommendation for my personal blog. Would greatly appreciate your input

trajan0x commented:
Absolutely. Wanna dm me somewhere?

oliverqx commented:
can i get your discord?

oliverqx commented:
@trajan0x mine is @olxverqx, or oliver@quixchan.com
